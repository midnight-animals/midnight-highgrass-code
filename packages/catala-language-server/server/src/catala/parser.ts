import * as fs from 'fs';
import * as path from 'path';
import * as Parser from 'tree-sitter';
// import * as Catala from '/home/user/dev/projects/tree-sitter/tree-sitter-catala/bindings/node/index.js';
import * as Catala from '../catala/tree-sitter/bindings/node';
import {
  CatalaFileParsed,
  CatalaGrammarTypes,
  CatalaSyntaxNode,
} from '../shared/CatalaTypes';
import {
  getAscendantNodeTypesPath,
  getAscendantTypesUntilSourceFile,
  getClosestAscendantOfTypes,
  getTypeFromClosestAscendantOfTypes,
} from './tree-sitter-focus/tree-sitter-modules';
import {
  areArraysEqual,
  arrayAlreadyIncludesArray,
} from '../shared/arrayUtils';

export async function parse(text: string): Promise<CatalaFileParsed> {
  const parser = new Parser();
  parser.setLanguage(Catala);
  const tree = parser.parse(text);
  const catalaFile: CatalaFileParsed = {
    error: [],
    constructor_name: [],
    enum_struct_name: [],
    field_name: [],
    scope_name: [],
    variable: [],
    verb_block: [],
    tree,
  };
  walkRecursive(tree.rootNode as CatalaSyntaxNode, catalaFile);
  return catalaFile;
}

function walkRecursive(
  node: CatalaSyntaxNode | undefined,
  catalaFile: CatalaFileParsed
) {
  if (!node) return;
  const catalaSyntaxNode: CatalaSyntaxNode = Object.assign(node, {
    enclosingBlockType: 'ALL',
  });

  // ['enum_struct_name', 'scope_decl_item', 'scope_decl']
  if (node.type === 'ERROR') {
    catalaFile.error.push(catalaSyntaxNode);
  }

  if (node.type === 'constructor_name') {
    const collectedTypes: CatalaGrammarTypes[] = [];
    getAscendantTypesUntilSourceFile(node, collectedTypes);
    if (collectedTypes.length > 2) {
      /* prettier-ignore */ console.log('%c------------------------------------------------------------------------------------------', `background: ${'darkblue'}`);
      node.type; /*?*/
      collectedTypes; /*?*/
    }

    const parentTypes: CatalaGrammarTypes[] = ['enum_decl_item'];
    const grandParentTypes: CatalaGrammarTypes[] = ['enum_decl'];
    const ascendantTypesPath = getAscendantNodeTypesPath(
      node,
      parentTypes,
      grandParentTypes
    );
    catalaSyntaxNode.ascendantTypesPath = ascendantTypesPath;
    catalaFile.constructor_name.push(catalaSyntaxNode);
  } else if (node.type === 'enum_struct_name') {
    const collectedTypes: CatalaGrammarTypes[] = [];
    getAscendantTypesUntilSourceFile(node, collectedTypes);
    if (collectedTypes.length > 2) {
      /* prettier-ignore */ console.log('%c------------------------------------------------------------------------------------------', `background: ${'darkblue'}`);
      node.type; /*?*/
      collectedTypes; /*?*/
    }

    const parentTypes: CatalaGrammarTypes[] = [
      'definition',
      'enum_decl',
      'struct_decl',
      'scope_decl_item',
    ];
    const grandParentTypes: CatalaGrammarTypes[] = ['scope', 'scope_decl'];
    const ascendantTypesPath = getAscendantNodeTypesPath(
      node,
      parentTypes,
      grandParentTypes
    );
    catalaSyntaxNode.ascendantTypesPath = ascendantTypesPath;
    catalaFile.enum_struct_name.push(catalaSyntaxNode);
  } else if (node.type === 'field_name') {
    const collectedTypes: CatalaGrammarTypes[] = [];
    getAscendantTypesUntilSourceFile(node, collectedTypes);
    if (collectedTypes.length > 2) {
      /* prettier-ignore */ console.log('%c------------------------------------------------------------------------------------------', `background: ${'darkblue'}`);
      node.type; /*?*/
      collectedTypes; /*?*/
    }

    const parentTypes: CatalaGrammarTypes[] = [
      'definition',
      'struct_decl',
      'struct_decl_item',
    ];
    const grandParentTypes: CatalaGrammarTypes[] = [
      'scope',
      'scope_decl',
      'struct_decl',
    ];
    const ascendantTypesPath = getAscendantNodeTypesPath(
      node,
      parentTypes,
      grandParentTypes
    );
    catalaSyntaxNode.ascendantTypesPath = ascendantTypesPath;
    catalaFile.field_name.push(catalaSyntaxNode);
  } else if (node.type === 'scope_name') {
    const collectedTypes: CatalaGrammarTypes[] = [];
    getAscendantTypesUntilSourceFile(node, collectedTypes);
    if (collectedTypes.length > 2) {
      /* prettier-ignore */ console.log('%c------------------------------------------------------------------------------------------', `background: ${'darkblue'}`);
      node.type; /*?*/
      collectedTypes; /*?*/
    }

    const parentTypes: CatalaGrammarTypes[] = [
      'scope_decl_item',
      'scope',
      'scope_decl',
    ];
    const grandParentTypes: CatalaGrammarTypes[] = ['scope', 'scope_decl'];
    const ascendantTypesPath = getAscendantNodeTypesPath(
      node,
      parentTypes,
      grandParentTypes
    );
    catalaSyntaxNode.ascendantTypesPath = ascendantTypesPath;
    catalaFile.scope_name.push(catalaSyntaxNode);
  } else if (node.type === 'variable') {
    const collectedTypes: CatalaGrammarTypes[] = [];
    getAscendantTypesUntilSourceFile(node, collectedTypes);
    /* prettier-ignore */ console.log('%c------------------------------------------------------------------------------------------', `background: ${'darkblue'}`);
    node.type; /*?*/
    collectedTypes; /*?*/

    const parentTypes: CatalaGrammarTypes[] = [
      'assertion',
      'definition',
      'rule',
      'scope_decl_item',
    ];
    const grandParentTypes: CatalaGrammarTypes[] = ['scope', 'scope_decl'];
    const ascendantTypesPath = getAscendantNodeTypesPath(
      node,
      parentTypes,
      grandParentTypes
    );

    catalaSyntaxNode.ascendantTypesPath = ascendantTypesPath;
    catalaFile.variable.push(catalaSyntaxNode);
  } else if (node.type === 'verb_block') {
    const collectedTypes: CatalaGrammarTypes[] = [];
    getAscendantTypesUntilSourceFile(node, collectedTypes);
    if (collectedTypes.length > 2) {
      /* prettier-ignore */ console.log('%c------------------------------------------------------------------------------------------', `background: ${'darkblue'}`);
      node.type; /*?*/
      collectedTypes; /*?*/
    }

    const parentTypes: CatalaGrammarTypes[] = ['definition', 'scope_decl_item'];
    const grandParentTypes: CatalaGrammarTypes[] = ['scope', 'scope_decl'];
    const ascendantTypesPath = getAscendantNodeTypesPath(
      node,
      parentTypes,
      grandParentTypes
    );

    catalaSyntaxNode.ascendantTypesPath = ascendantTypesPath;
    catalaFile.variable.push(catalaSyntaxNode);
  }

  if (node.childCount === 0) return;

  node.children.forEach((child) => {
    walkRecursive(child, catalaFile);
  });
}

const content = fs.readFileSync(
  path.resolve(__dirname, '../..', 'tests/fixtures/cli.catala_en'),
  'utf-8'
);

async function main() {
  const result = await parse(content).catch((err) => console.error(err));
  const { tree, ...parsed } = result as any;

  const map: Record<string, CatalaGrammarTypes[][]> = {};
  Object.entries(parsed).forEach(([key, value]) => {
    map[key] = [];
    parsed[key].forEach((node: CatalaSyntaxNode) => {
      if (arrayAlreadyIncludesArray(map[key], node.ascendantTypesPath)) return;

      if (node.ascendantTypesPath.length === 0) {
        node.ascendantTypesPath; /*?*/
        node.parent.parent.text; /*?*/
        node.text; /*?*/
      }

      map[key].push(node.ascendantTypesPath);
    });
  });

  map; /*?*/
  // parsed.enum_struct_name[0].ascendantTypesPath; /*?*/
}
main();
