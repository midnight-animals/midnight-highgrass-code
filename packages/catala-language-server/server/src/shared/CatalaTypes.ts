import * as Parser from 'tree-sitter';
import * as GrammarJson from '../catala/tree-sitter/src/grammar.json';

export type Id = number;
export type CatalaLanguages = 'en' | 'fr' | 'pl';

export type CatalaGrammarTypes = keyof typeof GrammarJson['rules'] | 'ERROR';
export interface CatalaSyntaxNode extends Parser.SyntaxNode {
  type: CatalaGrammarTypes;
  children: CatalaSyntaxNode[];
  nextSibling: CatalaSyntaxNode;
  parent: CatalaSyntaxNode;
  /**
   *
   * Note, that this is not the immediate parent like `parent`,
   * instead, this is the parent
   */
  enclosingBlockType: CatalaGrammarTypes | undefined;
  ascendantTypesPath: CatalaGrammarTypes[];
}

export enum DeclarationType {
  Scope = 'Scope',
  /** TODO ISSUE-6FErb8ou Can I assume, that structs are unique? */
  Struct = 'Struct',
  StructField = 'StructField',
}

/**
 * declaration scope IncomeTaxComputation:
 *   input individual content Individual   <---
 */
export interface CatalaStructureField {
  id: Id;
  text: string;
  type: DeclarationType | string;
  scopeId?: Id;
  structId?: Id;
}

export interface CatalaDeclaration {
  id: Id;
  text: string;
  type: DeclarationType;
  startPosition: Parser.SyntaxNode['startPosition'];
  endPosition: Parser.SyntaxNode['endPosition'];
}

export interface CatalaFileParsed {
  error: CatalaSyntaxNode[];
  enum_struct_name: CatalaSyntaxNode[];
  field_name: CatalaSyntaxNode[];
  scope_name: CatalaSyntaxNode[];
  variable: CatalaSyntaxNode[];
  verb_block: CatalaSyntaxNode[];
  tree: Parser.Tree;
}

export interface CatalaFile extends CatalaFileParsed {
  /** after going through UriUtils.toSysPath */
  uri: string;
}
