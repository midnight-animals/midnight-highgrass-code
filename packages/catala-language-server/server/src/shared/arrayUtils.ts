/* eslint-disable @typescript-eslint/no-unnecessary-type-constraint */

export function mergeArrays<T extends any>(arr1: T[], arr2: T[]): T[] {
  return Array.from(new Set([...(arr1 ?? []), ...(arr2 ?? [])]));
}

export function intersectArrays<T extends any>(arr1: T[], arr2: T[]): T[] {
  return arr1.filter((value) => arr2.includes(value));
}

/**
 * [ [ 'scope_decl_item', 'scope_decl' ],
 *   [ 'scope_decl_item', 'scope_decl' ] ],
 */
export function arrayAlreadyIncludesArray<T extends any>(
  arr: T[][],
  element: T[]
): boolean {
  const result = arr.find((item, index, self) => {
    const isSame = JSON.stringify(item) === JSON.stringify(element);
    return isSame;
    const stringifiedValue = JSON.stringify(item);
    return (
      index === self.findIndex((v) => JSON.stringify(v) === stringifiedValue)
    );
  });
  return !!result;
}

export function areArraysEqual<T extends any>(arr1: T[], arr2: T[]): boolean {
  return (
    arr1.length === arr2.length && arr1.every((value) => arr2.includes(value))
  );
}
