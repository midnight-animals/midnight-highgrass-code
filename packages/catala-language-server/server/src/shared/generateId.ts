let counter = 0;
export function generateId(): string {
  let mathRnd = Math.random().toString(25).slice(2);
  // let newId = Date.now().toString(25);
  mathRnd += counter++;

  return mathRnd;
}
