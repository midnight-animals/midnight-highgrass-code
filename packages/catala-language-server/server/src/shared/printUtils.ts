import { CatalaFileParsed } from './CatalaTypes';

export function printTree(treeInput: CatalaFileParsed): void {
  const { tree, ...rest } = treeInput;
  // console.log(JSON.stringify(rest, null, 2));
  rest; /*?*/
}
