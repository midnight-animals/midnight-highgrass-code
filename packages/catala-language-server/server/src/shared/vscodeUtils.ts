import { SyntaxNode } from 'tree-sitter';
import { pathToFileURL } from 'url';
import { LocationLink, Range, Position } from 'vscode-languageserver';
import { CatalaFile } from './CatalaTypes';

export function convertTreeSitterNodePositionToLocationLink(
  resultNode: SyntaxNode | undefined,
  targetFile: CatalaFile
): LocationLink {
  const range = convertNodeToRange(resultNode);
  const result = LocationLink.create(
    pathToFileURL(targetFile?.uri ?? '').toString(),
    range,
    range
  );
  return result;
}

export function convertNodeToRange(resultNode: SyntaxNode | undefined) {
  const result = Range.create(
    Position.create(
      resultNode?.startPosition.row ?? 0,
      resultNode?.startPosition.column ?? 0
    ),
    Position.create(
      resultNode?.endPosition.row ?? 0,
      resultNode?.endPosition.column ?? 0
    )
  );
  return result;
}
