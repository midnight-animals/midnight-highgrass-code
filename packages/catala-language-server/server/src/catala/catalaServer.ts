import {
  Connection,
  Diagnostic,
  DiagnosticSeverity,
  Position,
  TextDocuments,
} from 'vscode-languageserver';
import { TextDocument } from 'vscode-languageserver-textdocument';
import { parse } from './parser';
import { CatalaFile } from '../shared/CatalaTypes';
import { UriUtils } from '../shared/uri-utils';
import { onCompletion } from './capabilities/onCompletion';
import { onDefinition } from './capabilities/onDefinition';
import { convertNodeToRange } from '../shared/vscodeUtils';
import { closestAscendantOfType } from './tree-sitter-focus/tree-sitter-modules';
import { getCatalaCliOutput } from './modules/getCatalaCliOutput';

export class CatalaServer {
  private catalaFiles: CatalaFile[] = [];

  constructor(
    private connection: Connection,
    private documents: TextDocuments<TextDocument>
  ) {}

  public async validateTextDocument(textDocument: TextDocument) {
    console.log('[INFO] Validating text document...');
    if (this.catalaFiles.length === 0) return Promise.resolve([]);

    const targetFile = this.catalaFiles.find(
      (f) => f.uri === UriUtils.toSysPath(textDocument.uri)
    );
    if (!targetFile) return Promise.resolve([]);
    if (!targetFile.error.length) return Promise.resolve([]);

    const diagnostics: Diagnostic[] = [];

    const cliOutput = await getCatalaCliOutput(targetFile.tree.rootNode.text);
    if (!cliOutput?.range) return Promise.resolve([]);
    const message = cliOutput.text;
    console.error(message, ' at\n', cliOutput.range);
    const diagnostic: Diagnostic = {
      severity: DiagnosticSeverity.Error,
      range: cliOutput.range,
      source: 'catala-language-server',
      message,
    };
    diagnostics.push(diagnostic);

    return Promise.resolve(diagnostics);
  }

  public async onInitialized() {
    this.catalaFiles = [];
    // TODO ISSUE-JcuPKubs
    const extensionSettings = await this.connection.workspace.getConfiguration({
      section: '',
    });

    // parse current document
    for await (const document of this.documents.all()) {
      const content = document.getText();
      try {
        const parsed = await parse(content);
        const uri = UriUtils.toSysPath(document.uri);
        const file: CatalaFile = {
          ...parsed,
          uri,
        };
        this.catalaFiles.push(file);
      } catch (error) {
        console.error(error);
      }
    }

    // const logMe = JSON.stringify(this.catalaFiles[0], null, 2);
  }

  public onCompletion() {
    return onCompletion(this.catalaFiles);
  }

  public async onDefinition(document: TextDocument, position: Position) {
    const normalizedUri = UriUtils.toSysPath(document.uri);
    const targetFile = this.catalaFiles.find(
      (file) => file.uri === normalizedUri
    );
    if (!targetFile) return [];

    const definitions = onDefinition(targetFile, position);
    return definitions;
  }
}
