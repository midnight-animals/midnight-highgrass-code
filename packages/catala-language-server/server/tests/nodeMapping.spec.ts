import { parse } from '../src/catala/parser';
import {
  CatalaGrammarTypes,
  CatalaSyntaxNode,
} from '../src/shared/CatalaTypes';
import { arrayAlreadyIncludesArray } from '../src/shared/arrayUtils';
import { readCatalaFile } from './test-helpers/readCatalaTestFile';

describe('Node Mapping', () => {
  it('Node Mapping', async () => {
    const content = readCatalaFile('allSyntax');

    const result = await parse(content).catch((err) => console.error(err));
    const { tree, ...parsed } = result as any;

    const map: Record<string, CatalaGrammarTypes[][]> = {};
    Object.entries(parsed).forEach(([key, value]) => {
      map[key] = [];
      parsed[key].forEach((node: CatalaSyntaxNode) => {
        if (arrayAlreadyIncludesArray(map[key], node.ascendantTypesPath))
          return;

        map[key].push(node.ascendantTypesPath);
      });
    });

    expect(map).toMatchSnapshot();
  });
});
