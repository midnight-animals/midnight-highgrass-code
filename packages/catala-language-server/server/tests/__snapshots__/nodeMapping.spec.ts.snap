// Jest Snapshot v1, https://goo.gl/fbAQLP




/////////////////////////////////////
/////////////////////////////////////

refac file parser and compare against this snapshot
```bash
    const collectedTypes = [];
    getAscendantTypesUntilSourceFile(node, collectedTypes);
    if (collectedTypes.length > 2) {
      /* prettier-ignore */ console.log('%c------------------------------------------------------------------------------------------', `background: ${'darkblue'}`);
      node.type; /*?*/
      collectedTypes; /*?*/
    }
```

/////////////////////////////////////
/////////////////////////////////////

exports[`Node Mapping Node Mapping 1`] = `
{
  "constructor_name": [
    [
      "enum_decl_item",
      "enum_decl",
    ],
  ],
  "enum_struct_name": [
    [
      "struct_decl",
    ],
    [
      "enum_decl",
    ],
    [
      "scope_decl_item",
      "scope_decl",
    ],
    [
      "definition",
      "scope",
    ],
  ],
  "error": [],
  "field_name": [
    [
      "struct_decl_item",
      "struct_decl",
    ],
    [
      "definition",
      "scope",
    ],
  ],
  "scope_name": [
    [
      "scope_decl",
    ],
    [
      "scope",
    ],
    [
      "scope_decl_item",
      "scope_decl",
    ],
  ],
  "variable": [
    [
      "scope_decl_item",
      "scope_decl",
    ],
    [
      "definition",
      "scope",
    ],
    [
      "rule",
      "scope",
    ],
    [],
  ],
  "verb_block": [],
}
`;
